std/src/storage/mod.rs

```diff
@@ -0,0 +1,587 @@
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+{
+    pub fn new(key: S::Key, data: T) -> Self {
+        let channel = S::subscribe::<T>(&key);
+        Self {
+            entry: StorageEntry::new(key, data),
+            channel,
+        }
+    }
+
+    /// Gets the channel to subscribe to updates to the underlying storage
+    pub fn channel(&self) -> &Receiver<StorageChannelPayload> {
+        &self.channel
+    }
+
+    /// Creates a hook that will update the state when the underlying storage changes
+    pub fn subscribe_to_storage(&self) {
+        let storage_entry_signal = *self.data();
+        let channel = self.channel.clone();
+        spawn(async move {
+            to_owned![channel, storage_entry_signal];
+            loop {
+                // Wait for an update to the channel
+                if channel.changed().await.is_ok() {
+                    // Retrieve the latest value from the channel, mark it as read, and update the state
+                    let payload = channel.borrow_and_update();
+                    *storage_entry_signal.write() = payload
+                        .data
+                        .downcast_ref::<T>()
+                        .expect("Type mismatch with storage entry")
+                        .clone();
+                }
+            }
+        });
+    }
+}
+
+impl<S, T> StorageEntryTrait<S, T> for SyncedStorageEntry<S, T>
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+{
+    fn save(&self) {
+        //  We want to save in the following conditions
+        //      - The value from the channel is different from the current value
+        //      - The value from the channel could not be determined, likely because it hasn't been set yet
+        if let Some(payload) = self.channel.borrow().data.downcast_ref::<T>() {
+            if *self.entry.data.read() == *payload {
+                return;
+            }
+        }
+        self.entry.save();
+    }
+
+    fn update(&mut self) {
+        self.entry.update();
+    }
+
+    fn key(&self) -> &S::Key {
+        self.entry.key()
+    }
+
+    fn data(&self) -> &Signal<T> {
+        &self.entry.data
+    }
+}
+
+/// A storage entry that can be used to store data across application reloads. It optionally provides a channel to subscribe to updates to the underlying storage.
+#[derive(Clone)]
+pub struct StorageEntry<
+    S: StorageBacking,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + 'static,
+> {
+    /// The key used to store the data in storage
+    pub(crate) key: S::Key,
+    /// A signal that can be used to read and modify the state
+    pub(crate) data: Signal<T>,
+    /// An optional channel to subscribe to updates to the underlying storage
+    /// A lock to prevent multiple saves from happening at the same time
+    storage_save_lock: Arc<Mutex<()>>, // TODO: probably unnecessary
```

Comment:
This indeed looks unnecessary:
- fs: No need here, already has a lock internally
- memory: maybe actually add a lock here, Maybe use Arc<Mutex> instead of Rc<RefCell> ?
- web: No need here, I am pretty sure LocalStorage/SessionStorage have some kind of semaphore/queue inside, it's js at the end of the day hehe

Reply:
The memory storage already has a internal lock. I'm supprized clippy doesn't have a lint for unit mutexes